File: python/template-method/DataProcessor.py
from abc import ABC, abstractmethod

# Classe abstrata definindo o template
class DataProcessor(ABC):
    # Template method
    def process_data(self):
        self.read_data()
        self.process_data_implementation()
        self.save_data()

    # Passos concretos
    def read_data(self):
        print("Lendo dados...")

    def save_data(self):
        print("Salvando dados...")

    # Passo abstrato que será implementado pelas subclasses
    @abstractmethod
    def process_data_implementation(self):
        pass

File: java/template-method/template-method/src/main/java/edu/uea/DataProcessor.java
package edu.uea;

// Classe abstrata definindo o template
abstract class DataProcessor {
    // Template method
    public final void processData() {
        readData();
        processDataImplementation();
        saveData();
    }

    // Passos concretos
    protected void readData() {
        System.out.println("Lendo dados...");
    }

    protected void saveData() {
        System.out.println("Salvando dados...");
    }

    // Passo abstrato que será implementado pelas subclasses
    protected abstract void processDataImplementation();
}